Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/openSAML/HOKSAMLAssertionBuilder.java

```diff
@@ -61,667 +55,701 @@
+			final PublicKey publicKey) throws SAMLAssertionBuilderException {
+		Element assertionElement = null;
+		try {
+			final Signature signature = OpenSAML2ComponentBuilder.getInstance().createSignature(certificate, privateKey,
+					publicKey);
+			final SamlAssertionWrapper wrapper = new SamlAssertionWrapper(assertion);
+
+			wrapper.setSignature(signature, SignatureConstants.ALGO_ID_DIGEST_SHA1);
+			final MarshallerFactory marshallerFactory = XMLObjectProviderRegistrySupport.getMarshallerFactory();
+			final Marshaller marshaller = marshallerFactory.getMarshaller(wrapper.getSamlObject());
+			assertionElement = marshaller.marshall(wrapper.getSamlObject());
+			Signer.signObject(signature);
+		} catch (final SignatureException | org.apache.wss4j.common.ext.WSSecurityException
+				| org.opensaml.core.xml.io.MarshallingException e) {
+			LOG.error("Unable to sign: {}", e.getLocalizedMessage());
+			throw new SAMLAssertionBuilderException(e.getLocalizedMessage(), e);
+		}
+		return assertionElement;
+
+	}
+
+	protected Issuer createIssuer(final CallbackProperties properties) {
+		Issuer issuer;
+
+		final String format = properties.getAssertionIssuerFormat();
+		if (format != null) {
+			LOG.debug("Setting Assertion Issuer format to: {}", format);
+			final String sIssuer = properties.getIssuer();
+
+			LOG.debug("Setting Assertion Issuer to: {}", sIssuer);
+
+			if (isValidNameidFormat(format)) {
+				issuer = OpenSAML2ComponentBuilder.getInstance().createIssuer(format, sIssuer);
+			} else {
+				LOG.debug("Not in valid listing of formats: Using default issuer");
+				issuer = OpenSAML2ComponentBuilder.getInstance().createDefaultIssuer();
+			}
+		} else {
+			LOG.debug("Assertion issuer not defined: Using default issuer");
+			issuer = OpenSAML2ComponentBuilder.getInstance().createDefaultIssuer();
+		}
+		return issuer;
+	}
+
+	/**
+	 * @param PSApk
+	 * @param certificate
+	 * @return
+	 * @throws Exception
+	 */
+	protected Subject createSubject(final CallbackProperties properties, final X509Certificate certificate,
+			final PublicKey publicKey) throws SAMLComponentBuilderException {
+		String x509Name = properties.getUsername();
+
+		if ((NullChecker.isNullish(x509Name) || !checkDistinguishedName(x509Name)) && null != certificate
+				&& null != certificate.getSubjectDN()) {
+
+			x509Name = certificate.getSubjectDN().getName();
+
+		}
+		return createSubject(x509Name,  publicKey);
+	}
+
+	/**
+	 *
+	 * @param userName
+	 * @return boolean
+	 */
+	private static boolean checkDistinguishedName(final String userName) {
+		Name name = null;
+		try {
+			name = new LdapName(userName);
+		} catch (final Exception e) {
+			LOG.warn("Not a Valid Distinguished Name, setting the value from Certificate: {}", e.getLocalizedMessage());
+			LOG.trace("Exception checking DN: {}", e.getLocalizedMessage(), e);
+		}
+		return name == null ? false : true;
```

Comment:
Please revert the old code base for this method.  The old one is easier to understand and don't encounter additional check for performance issue.